DS/DS4/GuestHouse.java

```diff
@@ -0,0 +1,2575 @@
+import java.io.BufferedReader;
+import java.io.InputStreamReader;
+import java.util.ArrayList;
+import java.util.Enumeration;
+import java.util.Hashtable;
+import java.util.List;
+import java.util.Scanner;
+class RoomAlotement {
+	public List<String> showRoomStatus(Enumeration<Integer> roomNumbers,Hashtable<Integer, String> rooms,int roomNumber) {
+		List< String> roomStatus=new ArrayList<String>();
+		roomNumbers = rooms.keys();	// room numbers hold all the keys (room numbers) from hash table
+		while(roomNumbers.hasMoreElements()) {
+	         roomNumber = roomNumbers.nextElement();
+	         roomStatus.add("Room "+(roomNumber) + " : " +rooms.get(roomNumber));
+	           // show the status of each room
+		}
+		return roomStatus;
+	}
+	public String allotRoom(Hashtable<Integer, String>rooms,int age,String name,int numberOfRooms) {
+		int i,count=0,hashCode;
+		String result="";
+		boolean flag = true;
+		hashCode = age % numberOfRooms;	// if age is greater than 18 than generate hash code
+		for(i = hashCode; i <= (numberOfRooms - 1);) {   // continue the loop until house is empty
+			if(rooms.get(i).equals("")) {   // if room is empty
+				rooms.put(i, name); // than allocate that room to that user
+				System.out.println("Room number "+i+" is alloted to "+name);
+				flag = false;
+				break;
+			}
+			else {
+				i += 3;	// if room is full than increment i with 3 
+			}
+		}
+		i = 0;
+		while(flag) {   // if room not found
+			for(int j = i; j <= (numberOfRooms - 1); j++) {  // than check rooms status from starting
+				if(rooms.get(j).equals("")) {   // if empty than allocate
+					rooms.put(j, name);
+					result="Room number "+j+" is alloted to "+name;
+					flag = false;
+					break;
+				}
+				else {
+					j += 3;
+				}
+			}
+			i++;
+		}
+		count++;
+		return result;
+	}
+	public String getPerticularRoomStatus(int roomNumber,Hashtable<Integer, String>rooms) {
+		String result="";
+		if( rooms.get(roomNumber) != "" ) // if room is allocated someone than print that person name
+			result="Room "+roomNumber+" is alloted to "+rooms.get(roomNumber);
+		else							// otherwise print not alloted to any one
+			result = "Room is not alloted to anyone yet";
+		return result;
+	}
+	public String getPerticularPersonStatus(String name,Hashtable<Integer,String> rooms,Enumeration<Integer> roomNumbers,int roomNumber) {
```

Comment:
What is the advantage of having an Enumeration 'roomNumbers'. Can't we get them from the hash-table?